### PR TITLE
[#5074] feat(hadoop-catalog): Support GCS fileset. 

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

1.  Add a bundled jar for Hadoop GCS jar.
2. Support GCS in Hadoop catalog.

### Why are the changes needed?

Users highly demand Fileset for GCS storage.

Fix: #5074 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

Manually, please see: HadoopGCPCatalogIT

### CODE REVIEW

**Comment by yuqi1129** [comment]
This PR is a follow-up of https://github.com/apache/gravitino/pull/5020 and is not ready for review until #5020 is merged.

File: catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.TestInstance;
+
+@Tag("gravitino-docker-test")
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@Disabled(
```
Comment (xloya) [inline catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java:39]: I think if ITs are not possible without a valid account, then we can add some unit tests for GCS. As I know, Iceberg had already implemented this, you can refer this: https://github.com/apache/iceberg/blob/main/gcp/src/test/java/org/apache/iceberg/gcp/gcs/GCSFileIOTest.java. But I haven't verified the feasibility, please help confirm this.

File: build.gradle.kts [hunk 2]
```diff
@@ -764,7 +764,7 @@ tasks {
         !it.name.startsWith("integration-test") &&
         !it.name.startsWith("flink") &&
         !it.name.startsWith("trino-connector") &&
-        it.name != "hive-metastore-common"
+        it.name != "hive-metastore-common" && it.name != "gcs-bundle"
```
Comment (jerryshao) [inline build.gradle.kts:767]: gcp-bundle, not gcs

File: catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    "Disabled due to as we don't have a real GCP account to test. If you have a GCP account,"
+        + "please change the configuration(YOUR_KEY_FILE, YOUR_BUCKET) and enable this test.")
+public class HadoopGCPCatalogIT extends HadoopCatalogIT {
```
Comment (jerryshao) [inline catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java:42]: You need to clarify the difference between gcp and gcs.

File: bundles/gcs-bundle/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+
+plugins {
+  `maven-publish`
+  id("java")
+  alias(libs.plugins.shadow)
+}
+
+dependencies {
+  compileOnly(project(":catalogs:catalog-hadoop"))
```
Comment (jerryshao) [inline bundles/gcs-bundle/build.gradle.kts:28]: I guess `compileOnly` is not enough for gvfs.

File: catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  public static final String BUCKET_NAME = "YOUR_BUCKET";
+  public static final String SERVICE_ACCOUNT_FILE = "YOUR_KEY_FILE";
+
+  @BeforeAll
+  public void setup() throws IOException {
```
Comment (jerryshao) [inline catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java:48]: It is better you can also have gvfs test for gcs.

File: bundles/gcs-bundle/build.gradle.kts (hunk 4 above)
Comment (yuqi1129) [reply bundles/gcs-bundle/build.gradle.kts:28]: I have added this depend as `implementation` in module `filesysem-hadoop3`, so I believe it's unnecessary for `gcs-bundle` jar.

File: catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java]: added.

File: catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java (hunk 1 above)
Comment (yuqi1129) [reply catalogs/catalog-hadoop/src/test/java/org/apache/gravitino/catalog/hadoop/integration/test/HadoopGCPCatalogIT.java]: Let me check it.

File: bundles/gcs-bundle/build.gradle.kts (hunk 4 above)
Comment (jerryshao) [reply bundles/gcs-bundle/build.gradle.kts]: OK.

**Comment by jerryshao** [comment]
How does user use gcp-bundle with gvfs, can you please give me an example?

**Comment by yuqi1129** [comment]
> How does user use gcp-bundle with gvfs, can you please give me an example?

1. Add the following dependency:
```xml
        <dependency>
            <groupId>org.apache.hadoop</groupId>
            <artifactId>hadoop-common</artifactId>
            <version>3.1.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.gravitino</groupId>
            <artifactId>gcp-bundle</artifactId>
            <version>0.7.0-incubating-SNAPSHOT</version>
        </dependency>

        <dependency>
            <groupId>org.apache.gravitino</groupId>
            <artifactId>filesystem-hadoop3-runtime</artifactId>
            <version>0.7.0-incubating-SNAPSHOT</version>
        </dependency>
```

2.  The following is an example code.
```
  public static void main(String[] args) throws IOException {

    Configuration conf = new Configuration();
    conf.set("fs.gvfs.impl", "org.apache.gravitino.filesystem.hadoop.GravitinoVirtualFileSystem");
    conf.set("fs.AbstractFileSystem.gvfs.impl", "org.apache.gravitino.filesystem.hadoop.Gvfs");
    conf.set("fs.gvfs.impl.disable.cache", "true");
    conf.set("fs.gravitino.server.uri", "http://127.0.0.1:8090");
    conf.set("fs.gravitino.client.metalake", "gvfs_it_metalake_1fd37007");

    // Pass this configuration to the real file system
    conf.set("gravitino.bypass.fs.gs.auth.service.account.enable", "true");
    conf.set("gravitino.bypass.fs.gs.auth.service.account.json.keyfile", SERVICE_ACCOUNT_FILE);
    conf.set(FS_FILESYSTEM_PROVIDERS, "gcs");

    String gvfsPath = "gvfs://fileset/catalog_704cac87/schema_858f9d26/test_fileset_create_88eaed4e";
    Path path = new Path(gvfsPath);

    FileSystem f = path.getFileSystem(conf);

    System.out.println("fileSystem: " + f);

    String filePath = gvfsPath + "/test.txt";

    System.out.println(f.exists(new Path(filePath)));
  }

```

@jerryshao 
I have confirmed locally and all are as expected.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**